tests/components/test_upnp.py

```diff
@@ -0,0 +1,139 @@
+        yield
+
+
+@pytest.fixture(autouse=True)
+def mock_discovery():
+    """Mock discovery of upnp sensor."""
+    with patch('homeassistant.components.upnp.discovery'):
+        yield
+
+
+@asyncio.coroutine
+def test_setup_fail_if_no_ip(hass):
+    """Test setup fails if we can't find a local IP."""
+    with patch('homeassistant.components.upnp.get_local_ip',
+               return_value='127.0.0.1'):
+        result = yield from async_setup_component(hass, 'upnp', {
+            'upnp': {}
+        })
+
+    assert not result
+
+
+@asyncio.coroutine
+def test_setup_fail_if_cannot_select_igd(hass, mock_local_ip, mock_miniupnpc):
+    """Test setup fails if we can't find an UPnP IGD."""
+    mock_miniupnpc.selectigd.side_effect = Exception
+
+    result = yield from async_setup_component(hass, 'upnp', {
+        'upnp': {}
+    })
+
+    assert not result
+
+
+@asyncio.coroutine
+def test_setup_succeeds_if_specify_ip(hass, mock_miniupnpc):
+    """Test setup succeeds if we specify IP and can't find a local IP."""
+    with patch('homeassistant.components.upnp.get_local_ip',
+               return_value='127.0.0.1'):
+        result = yield from async_setup_component(hass, 'upnp', {
+            'upnp': {
+                'local_ip': '192.168.0.10'
+            }
+        })
+
+    assert result
+
+
+@asyncio.coroutine
+def test_no_config_maps_hass_local_to_remote_port(hass, mock_miniupnpc):
+    """Test by default we map local to remote port."""
+    result = yield from async_setup_component(hass, 'upnp', {
+        'upnp': {
+            'local_ip': '192.168.0.10'
+        }
+    })
+
+    assert result
+    assert len(mock_miniupnpc.addportmapping.mock_calls) == 1
+    external, _, host, internal, _, _ = \
+        mock_miniupnpc.addportmapping.mock_calls[0][1]
+    assert external == 8123
+    assert internal == 8123
+
+
+@asyncio.coroutine
+def test_map_hass_to_remote_port(hass, mock_miniupnpc):
+    """Test mapping hass to remote port."""
+    result = yield from async_setup_component(hass, 'upnp', {
+        'upnp': {
+            'local_ip': '192.168.0.10',
+            'ports': {
+                'hass': 1000
+            }
+        }
+    })
+
+    assert result
+    assert len(mock_miniupnpc.addportmapping.mock_calls) == 1
+    external, _, host, internal, _, _ = \
+        mock_miniupnpc.addportmapping.mock_calls[0][1]
+    assert external == 1000
+    assert internal == 8123
+
+
+@asyncio.coroutine
+def test_map_internal_to_remote_ports(hass, mock_miniupnpc):
+    """Test mapping local to remote ports."""
+    result = yield from async_setup_component(hass, 'upnp', {
+        'upnp': {
+            'local_ip': '192.168.0.10',
+            'ports': {
+                'hass': 1000,
+                1883: 3883
+            }
+        }
+    })
+
+    assert result
+    assert len(mock_miniupnpc.addportmapping.mock_calls) == 2
+    external, _, host, internal, _, _ = \
+        mock_miniupnpc.addportmapping.mock_calls[0][1]
```

Comment:
Can you be certain of the order of these calls? Is port's dict an ordered dict?

Reply:
It's not in the test but it is when it comes from yaml. Will fix the test